Processing stops because sessions of the DistributedLogStreamClient are closed

**Describe the bug**

Lately we observed in our long running test suite that the processing of one partition (eventually most of the partitions) stop processing.

![processing-stopped](https://user-images.githubusercontent.com/2758593/68862559-f5424c00-06ed-11ea-88ae-35f1a7df96d3.png)

We saw no broker restart nor leader change at this time.
The only thing we saw at this point in time was the reconfiguration of the DistributedLogPrimitive.
That is the reason why we assumed the session might be closed and the append was not happening afterwards.

If we take a look at the Heartbeat rate we see that at some point only the real heartbeat requests are send, no longer any append requests.

![heartbeats](https://user-images.githubusercontent.com/2758593/68862733-44887c80-06ee-11ea-9cf6-61a1cb83364c.png)

In the log we saw log statements like this:

```
I 2019-11-14T13:29:40.145523760Z zeebe-1 2019-11-14 13:29:40.144 [] [raft-partition-group-raft-atomix-13] ERROR io.zeebe.logstreams - Failed to append block with last event position 1099516985048, retry.
  zeebe-1
I 2019-11-14T13:29:40.145546797Z zeebe-1 java.util.concurrent.CompletionException: io.atomix.primitive.PrimitiveException$ClosedSession: session closed
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
	at io.atomix.utils.concurrent.Futures.lambda$null$3(Futures.java:159) ~[atomix-utils-3.2.0-20191112.114214-59.jar:?]
	at io.atomix.utils.concurrent.ThreadPoolContext.lambda$new$0(ThreadPoolContext.java:81) ~[atomix-utils-3.2.0-20191112.114214-59.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: io.atomix.primitive.PrimitiveException$ClosedSession: session closed
	at io.atomix.protocols.raft.session.impl.RaftSessionInvoker.close(RaftSessionInvoker.java:217) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.DefaultRaftSessionClient.lambda$null$0(DefaultRaftSessionClient.java:236) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionState.lambda$setState$0(RaftSessionState.java:221) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(Unknown Source) ~[?:?]
	at java.util.concurrent.CopyOnWriteArraySet.forEach(Unknown Source) ~[?:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionState.setState(RaftSessionState.java:221) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionInvoker$CommandAttempt.accept(RaftSessionInvoker.java:388) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionInvoker$CommandAttempt.accept(RaftSessionInvoker.java:327) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.complete(Unknown Source) ~[?:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionConnection.handleResponse(RaftSessionConnection.java:282) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionConnection.lambda$sendRequest$7(RaftSessionConnection.java:238) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(Unknown Source) ~[?:?]
	... 7 more
  zeebe-1
I 2019-11-14T13:29:40.146362510Z zeebe-1 2019-11-14 13:29:40.145 [] [raft-partition-group-raft-atomix-13] ERROR io.zeebe.logstreams - Failed to append block with last event position 1099516999056, retry.
  zeebe-1
I 2019-11-14T13:29:40.146404536Z zeebe-1 java.util.concurrent.CompletionException: io.atomix.primitive.PrimitiveException$ClosedSession: session closed
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
	at io.atomix.utils.concurrent.Futures.lambda$null$3(Futures.java:159) ~[atomix-utils-3.2.0-20191112.114214-59.jar:?]
	at io.atomix.utils.concurrent.ThreadPoolContext.lambda$new$0(ThreadPoolContext.java:81) ~[atomix-utils-3.2.0-20191112.114214-59.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: io.atomix.primitive.PrimitiveException$ClosedSession: session closed
	at io.atomix.protocols.raft.session.impl.RaftSessionInvoker.close(RaftSessionInvoker.java:217) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.DefaultRaftSessionClient.lambda$null$0(DefaultRaftSessionClient.java:236) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionState.lambda$setState$0(RaftSessionState.java:221) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(Unknown Source) ~[?:?]
	at java.util.concurrent.CopyOnWriteArraySet.forEach(Unknown Source) ~[?:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionState.setState(RaftSessionState.java:221) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionInvoker$CommandAttempt.accept(RaftSessionInvoker.java:388) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionInvoker$CommandAttempt.accept(RaftSessionInvoker.java:327) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.complete(Unknown Source) ~[?:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionConnection.handleResponse(RaftSessionConnection.java:282) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionConnection.lambda$sendRequest$7(RaftSessionConnection.java:238) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(Unknown Source) ~[?:?]
	... 7 more
  zeebe-1
I 2019-11-14T13:29:40.147062892Z zeebe-1 2019-11-14 13:29:40.146 [] [raft-partition-group-raft-atomix-13] ERROR io.zeebe.logstreams - Failed to append block with last event position 1099516999352, retry.
  zeebe-1
```

As you can see in the following code from the `LogStorageAppender`:

```java
  private void appendToPrimitive(byte[] bytesToAppend, long lastEventPosition) {
    actor.submit(
        () -> {
          distributedLog
              .asyncAppend(bytesToAppend, lastEventPosition)
              .whenComplete(
                  (appendPosition, error) -> {
                    if (error != null) {
                      LOG.error(
                          "Failed to append block with last event position {}, retry.",
                          lastEventPosition,
                          error);
                      appendToPrimitive(bytesToAppend, lastEventPosition);
                    } else {
                      actor.run(
                          () -> {
                            appendEntryLimiter.onCommit(lastEventPosition);
                            currentInFlightBytes -= bytesToAppend.length;
                            actor.run(this::peekBlock);
                          });
                    }
                  });
        });
  }
```

The append is retried on error, but `actor.submit` will throw an exception since it is not called in the `ActorThread` it self. This means the retry will fail.

But even if the retry will succeed we would have the problem that it might happen that events are then appended out of order. 

**Example**:

We append `1, 2, 3, 4` and the session close exception will happen for 2 and 3. Then it can happen that we have in the end `1, 4, 2, 3` as appended entries. This might be detected by the `DistributedLogPrimitive`, such that it does not append `2` and `3` to the log storage.

--- a/broker/src/test/java/io/zeebe/broker/transport/commandapi/CommandApiMessageHandlerTest.java
+++ b/broker/src/test/java/io/zeebe/broker/transport/commandapi/CommandApiMessageHandlerTest.java
@@ -13,6 +13,7 @@ import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyLong;
 import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.doReturn;
 import static org.mockito.Mockito.mock;
 
 import com.netflix.concurrency.limits.limit.SettableLimit;
@@ -89,7 +90,7 @@ public class CommandApiMessageHandlerTest {
   private LogStream logStream;
   private CommandApiMessageHandler messageHandler;
   private DistributedLogstreamService distributedLogImpl;
-  private RequestLimiter noneLimiter = new NoopRequestLimiter();
+  private final RequestLimiter noneLimiter = new NoopRequestLimiter();
 
   @Before
   public void setup() {
@@ -133,15 +134,18 @@ public class CommandApiMessageHandlerTest {
                       && arguments.length > 1
                       && arguments[0] != null
                       && arguments[1] != null) {
-                    final byte[] bytes = (byte[]) arguments[0];
-                    final long pos = (long) arguments[1];
+                    final long appendIndex = (long) arguments[0];
+                    final byte[] bytes = (byte[]) arguments[1];
+                    final long pos = (long) arguments[2];
                     return CompletableFuture.completedFuture(
-                        distributedLogImpl.append(nodeId, pos, bytes));
+                        distributedLogImpl.append(nodeId, appendIndex, pos, bytes));
                   }
                   return null;
                 })
         .when(mockDistLog)
-        .asyncAppend(any(byte[].class), anyLong());
+        .asyncAppend(anyLong(), any(byte[].class), anyLong());
+
+    doReturn(CompletableFuture.completedFuture(0L)).when(mockDistLog).getLastAppendIndex();
 
     serviceContainerRule
         .get()

--- a/engine/src/test/java/io/zeebe/engine/util/TestStreams.java
+++ b/engine/src/test/java/io/zeebe/engine/util/TestStreams.java
@@ -13,6 +13,7 @@ import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyInt;
 import static org.mockito.ArgumentMatchers.anyLong;
 import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.doReturn;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.when;
@@ -175,15 +176,17 @@ public class TestStreams {
                       && arguments.length > 1
                       && arguments[0] != null
                       && arguments[1] != null) {
-                    final byte[] bytes = (byte[]) arguments[0];
-                    final long pos = (long) arguments[1];
+                    final long appendIndex = (long) arguments[0];
+                    final byte[] bytes = (byte[]) arguments[1];
+                    final long pos = (long) arguments[2];
                     return CompletableFuture.completedFuture(
-                        distributedLogImpl.append(nodeId, pos, bytes));
+                        distributedLogImpl.append(nodeId, appendIndex, pos, bytes));
                   }
                   return null;
                 })
         .when(mockDistLog)
-        .asyncAppend(any(byte[].class), anyLong());
+        .asyncAppend(anyLong(), any(byte[].class), anyLong());
+    doReturn(CompletableFuture.completedFuture(0L)).when(mockDistLog).getLastAppendIndex();
 
     serviceContainer
         .createService(distributedLogPartitionServiceName(name), () -> mockDistLog)

--- a/logstreams/src/main/java/io/zeebe/distributedlog/AsyncDistributedLogstream.java
+++ b/logstreams/src/main/java/io/zeebe/distributedlog/AsyncDistributedLogstream.java
@@ -13,9 +13,11 @@ import java.util.concurrent.CompletableFuture;
 public interface AsyncDistributedLogstream extends AsyncPrimitive {
 
   CompletableFuture<Long> append(
-      String partition, String nodeId, long commitPosition, byte[] blockBuffer);
+      String partition, String nodeId, long appendIndex, long commitPosition, byte[] blockBuffer);
 
   CompletableFuture<Boolean> claimLeaderShip(String partition, String nodeId, long leaderTerm);
+
+  CompletableFuture<Long> lastAppendIndex(String partition);
 
   @Override
   DistributedLogstream sync();

--- a/logstreams/src/main/java/io/zeebe/distributedlog/DistributedLogstream.java
+++ b/logstreams/src/main/java/io/zeebe/distributedlog/DistributedLogstream.java
@@ -13,7 +13,11 @@ import java.util.concurrent.TimeoutException;
 
 public interface DistributedLogstream extends SyncPrimitive {
 
-  long append(String partition, String nodeId, long commitPosition, byte[] blockBuffer)
+  long append(
+      String partition, String nodeId, long appendIndex, long commitPosition, byte[] blockBuffer)
+      throws InterruptedException, ExecutionException, TimeoutException;
+
+  long lastAppendIndex(String partition)
       throws InterruptedException, ExecutionException, TimeoutException;
 
   @Override

--- a/logstreams/src/main/java/io/zeebe/distributedlog/DistributedLogstreamService.java
+++ b/logstreams/src/main/java/io/zeebe/distributedlog/DistributedLogstreamService.java
@@ -8,12 +8,17 @@
 package io.zeebe.distributedlog;
 
 import io.atomix.primitive.operation.Command;
+import io.atomix.primitive.operation.Query;
 
 public interface DistributedLogstreamService {
 
   /* Returns a positive value if the append was successful */
   @Command
-  long append(String nodeId, long commitPosition, byte[] blockBuffer);
+  long append(String nodeId, long appendIndex, long commitPosition, byte[] blockBuffer);
+
+  /* Returns the last successful appended index by the primitive. */
+  @Query
+  long lastAppendIndex();
 
   /* Return true if operation was successful */
   @Command

--- a/logstreams/src/main/java/io/zeebe/distributedlog/impl/BlockingDistributedLogstream.java
+++ b/logstreams/src/main/java/io/zeebe/distributedlog/impl/BlockingDistributedLogstream.java
@@ -28,11 +28,18 @@ public class BlockingDistributedLogstream extends Synchronous<AsyncDistributedLo
   }
 
   @Override
-  public long append(String partition, String nodeId, long commitPosition, byte[] blockBuffer)
+  public long append(
+      String partition, String nodeId, long appendIndex, long commitPosition, byte[] blockBuffer)
       throws InterruptedException, ExecutionException, TimeoutException {
     return distributedLogstreamProxy
-        .append(partition, nodeId, commitPosition, blockBuffer)
+        .append(partition, nodeId, appendIndex, commitPosition, blockBuffer)
         .get(timeout, TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public long lastAppendIndex(String partition)
+      throws InterruptedException, ExecutionException, TimeoutException {
+    return distributedLogstreamProxy.lastAppendIndex(partition).get(timeout, TimeUnit.MILLISECONDS);
   }
 
   @Override

--- a/logstreams/src/main/java/io/zeebe/distributedlog/impl/DefaultDistributedLogstreamService.java
+++ b/logstreams/src/main/java/io/zeebe/distributedlog/impl/DefaultDistributedLogstreamService.java
@@ -48,21 +48,23 @@ public class DefaultDistributedLogstreamService
     implements DistributedLogstreamService, LogReplicationAppender {
 
   private static final Logger LOG =
-      LoggerFactory.getLogger(DefaultDistributedLogstreamService.class);
-
+      LoggerFactory.getLogger(DefaultDistributedLogstreamService.class);;
   private LogStream logStream;
   private String logName;
   private int partitionId;
   private String currentLeader;
   private long currentLeaderTerm = -1;
   private long lastPosition;
+  private long lastAppendIndex;
   private ServiceContainer serviceContainer;
   private String localMemberId;
   private Logger logger;
+  private long currentAppendIndex;
 
   public DefaultDistributedLogstreamService() {
     super(DistributedLogstreamType.instance(), DistributedLogstreamClient.class);
     lastPosition = -1;
+    lastAppendIndex = 0;
   }
 
   @Override
@@ -165,7 +167,8 @@ public class DefaultDistributedLogstreamService
   }
 
   @Override
-  public long append(String nodeId, long commitPosition, byte[] blockBuffer) {
+  /** append from the primitive */
+  public long append(String nodeId, long appendIndex, long commitPosition, byte[] blockBuffer) {
     // Assumption: first append is always called after claim leadership. So currentLeader is not
     // null. Assumption is also valid during restart.
     if (!currentLeader.equals(nodeId)) {
@@ -175,7 +178,22 @@ public class DefaultDistributedLogstreamService
       // error code;
     }
 
+    // if the lastAppendIndex + 1 is less then the appendIndex, we missed some events
+    if (lastAppendIndex + 1 < appendIndex) {
+      logger.trace(
+          "We received an event out of order. Last append index {} and try to append index {}, will reject this event.",
+          lastAppendIndex,
+          appendIndex);
+      return -lastAppendIndex;
+    }
+
+    currentAppendIndex = appendIndex;
     return append(commitPosition, blockBuffer);
+  }
+
+  @Override
+  public long lastAppendIndex() {
+    return lastAppendIndex;
   }
 
   @Override
@@ -194,12 +212,20 @@ public class DefaultDistributedLogstreamService
     return false;
   }
 
+  // this is called by log replication and primitive append
   @Override
   public long append(long commitPosition, byte[] blockBuffer) {
     if (commitPosition <= lastPosition) {
       // This case can happen due to raft-replay or when appender retries due to timeout or other
       // exceptions.
-      logger.trace("Rejecting append request at position {}", commitPosition);
+      logger.trace(
+          "Rejecting append request at position {} last position was {}",
+          commitPosition,
+          lastPosition);
+
+      if (lastAppendIndex <= currentAppendIndex) {
+        lastAppendIndex = currentAppendIndex;
+      }
       return 1; // Assume the append was successful because event was previously appended.
     }
 
@@ -209,6 +235,12 @@ public class DefaultDistributedLogstreamService
       return -1;
     }
 
+    final long lastAppendResult = appendToLogstorage(commitPosition, blockBuffer);
+    lastAppendIndex = currentAppendIndex;
+    return lastAppendResult;
+  }
+
+  private long appendToLogstorage(long commitPosition, byte[] blockBuffer) {
     try {
       final ByteBuffer buffer = ByteBuffer.wrap(blockBuffer);
       final long appendResult = logStream.append(commitPosition, buffer);
@@ -236,17 +268,25 @@ public class DefaultDistributedLogstreamService
     // backup snapshot, but not appended to logStorage may be lost. So, if this node is away for a
     // while and tries to recover with backup received from other nodes, there will be missing
     // entries in the logStorage.
-    logger.info("Backup log {} at position {}", logName, lastPosition);
+    logger.info(
+        "Backup log {} at position {}, last append index {}",
+        logName,
+        lastPosition,
+        lastAppendIndex);
 
     // Backup in-memory states
     backupOutput.writeLong(lastPosition);
+    backupOutput.writeLong(lastAppendIndex);
     backupOutput.writeString(currentLeader);
     backupOutput.writeLong(currentLeaderTerm);
   }
 
   @Override
   public void restore(BackupInput backupInput) {
+    LOG.debug("Restoring primitive of partition {}", partitionId);
     final long backupPosition = backupInput.readLong();
+    lastAppendIndex = backupInput.readLong();
+    currentAppendIndex = lastAppendIndex;
 
     if (lastPosition < backupPosition) {
       LogstreamConfig.startRestore(localMemberId, partitionId);
@@ -273,7 +313,10 @@ public class DefaultDistributedLogstreamService
       LogstreamConfig.completeRestore(localMemberId, partitionId);
     }
 
-    logger.debug("Restored local log to position {}", lastPosition);
+    logger.debug(
+        "Restored local log to position {} and last append index {}",
+        lastPosition,
+        lastAppendIndex);
     currentLeader = backupInput.readString();
     currentLeaderTerm = backupInput.readLong();
   }

--- a/logstreams/src/main/java/io/zeebe/distributedlog/impl/DistributedLogstreamPartition.java
+++ b/logstreams/src/main/java/io/zeebe/distributedlog/impl/DistributedLogstreamPartition.java
@@ -19,8 +19,6 @@ import io.zeebe.servicecontainer.ServiceStopContext;
 import io.zeebe.util.ZbLogger;
 import io.zeebe.util.sched.future.CompletableActorFuture;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.TimeoutException;
 import org.slf4j.Logger;
 
 public class DistributedLogstreamPartition implements Service<DistributedLogstreamPartition> {
@@ -46,13 +44,15 @@ public class DistributedLogstreamPartition implements Service<DistributedLogstre
     partitionName = DistributedLogstreamName.getPartitionKey(partitionId);
   }
 
-  public long append(byte[] blockBuffer, long commitPosition)
-      throws InterruptedException, ExecutionException, TimeoutException {
-    return distributedLog.append(partitionName, memberId, commitPosition, blockBuffer);
+  public CompletableFuture<Long> asyncAppend(
+      long appendIndex, byte[] blockBuffer, long commitPosition) {
+    return distributedLog
+        .async()
+        .append(partitionName, memberId, appendIndex, commitPosition, blockBuffer);
   }
 
-  public CompletableFuture<Long> asyncAppend(byte[] blockBuffer, long commitPosition) {
-    return distributedLog.async().append(partitionName, memberId, commitPosition, blockBuffer);
+  public CompletableFuture<Long> getLastAppendIndex() {
+    return distributedLog.async().lastAppendIndex(partitionName);
   }
 
   public CompletableFuture<Boolean> claimLeaderShip() {

--- a/logstreams/src/main/java/io/zeebe/distributedlog/impl/DistributedLogstreamProxy.java
+++ b/logstreams/src/main/java/io/zeebe/distributedlog/impl/DistributedLogstreamProxy.java
@@ -35,9 +35,14 @@ public class DistributedLogstreamProxy
 
   @Override
   public CompletableFuture<Long> append(
-      String partition, String nodeId, long commitPosition, byte[] buffer) {
+      String partition, String nodeId, long appendIndex, long commitPosition, byte[] buffer) {
     return getProxyClient()
-        .applyBy(partition, service -> service.append(nodeId, commitPosition, buffer));
+        .applyBy(partition, service -> service.append(nodeId, appendIndex, commitPosition, buffer));
+  }
+
+  @Override
+  public CompletableFuture<Long> lastAppendIndex(String partition) {
+    return getProxyClient().applyBy(partition, DistributedLogstreamService::lastAppendIndex);
   }
 
   @Override

--- a/logstreams/src/main/java/io/zeebe/logstreams/impl/LogStorageAppender.java
+++ b/logstreams/src/main/java/io/zeebe/logstreams/impl/LogStorageAppender.java
@@ -21,8 +21,10 @@ import io.zeebe.logstreams.impl.backpressure.AppenderVegasCfg;
 import io.zeebe.logstreams.impl.backpressure.BackpressureConstants;
 import io.zeebe.logstreams.impl.backpressure.NoopAppendLimiter;
 import io.zeebe.util.Environment;
+import io.zeebe.util.exception.UncheckedExecutionException;
 import io.zeebe.util.sched.Actor;
 import io.zeebe.util.sched.future.ActorFuture;
+import io.zeebe.util.sched.future.CompletableActorFuture;
 import java.nio.ByteBuffer;
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicBoolean;
@@ -49,6 +51,9 @@ public class LogStorageAppender extends Actor {
   private long currentInFlightBytes;
   private final Runnable peekedBlockHandler = this::appendBlock;
 
+  private long appendIndex = -1;
+  private final int partitionId;
+
   public LogStorageAppender(
       String name,
       DistributedLogstreamPartition distributedLog,
@@ -60,7 +65,7 @@ public class LogStorageAppender extends Actor {
     this.writeBufferSubscription = writeBufferSubscription;
     this.maxAppendBlockSize = maxBlockSize;
 
-    final int partitionId = distributedLog.getPartitionId();
+    partitionId = distributedLog.getPartitionId();
     appendBackpressureMetrics = new AppendBackpressureMetrics(partitionId);
 
     final boolean isBackpressureEnabled =
@@ -110,12 +115,14 @@ public class LogStorageAppender extends Actor {
     appendBackpressureMetrics.newEntryToAppend();
     if (appendEntryLimiter.tryAcquire(lastEventPosition)) {
       currentInFlightBytes += bytes;
-      appendToPrimitive(bytesToAppend, lastEventPosition);
+      ++appendIndex;
+      appendToPrimitive(appendIndex, bytesToAppend, lastEventPosition);
       blockPeek.markCompleted();
     } else {
       appendBackpressureMetrics.deferred();
       LOG.trace(
-          "Backpressure happens: in flight {} (in bytes {}) limit {}",
+          "Backpressure happens on partition {}: in flight {} (in bytes {}) limit {}",
+          partitionId,
           appendEntryLimiter.getInflight(),
           currentInFlightBytes,
           appendEntryLimiter.getLimit());
@@ -123,28 +130,38 @@ public class LogStorageAppender extends Actor {
     }
   }
 
-  private void appendToPrimitive(byte[] bytesToAppend, long lastEventPosition) {
-    actor.submit(
-        () -> {
-          distributedLog
-              .asyncAppend(bytesToAppend, lastEventPosition)
-              .whenComplete(
-                  (appendPosition, error) -> {
+  private void appendToPrimitive(long appendIndex, byte[] bytesToAppend, long lastEventPosition) {
+    distributedLog
+        .asyncAppend(appendIndex, bytesToAppend, lastEventPosition)
+        .whenComplete(
+            (appendPosition, error) -> {
+              // BE AWARE THIS IS CALLED BY A RAFT THREAD
+              actor.run(
+                  () -> {
                     if (error != null) {
-                      LOG.error(
-                          "Failed to append block with last event position {}, retry.",
+                      LOG.trace(
+                          "Append on partition {} failed with exception. Append index {} and event position {}. Retry...",
+                          partitionId,
+                          appendIndex,
+                          lastEventPosition,
+                          error);
+                      appendToPrimitive(appendIndex, bytesToAppend, lastEventPosition);
+                    } else if (appendPosition < 0) {
+                      LOG.trace(
+                          "Append on partition {} failed with negative result {}. Append index {} and event position {}. This normally occurs when an previous append failed and events have been received out of order. This is the recovery strategy to get events in order again. Retry...",
+                          partitionId,
+                          appendPosition,
+                          appendIndex,
                           lastEventPosition);
-                      appendToPrimitive(bytesToAppend, lastEventPosition);
+                      appendToPrimitive(appendIndex, bytesToAppend, lastEventPosition);
                     } else {
-                      actor.run(
-                          () -> {
-                            appendEntryLimiter.onCommit(lastEventPosition);
-                            currentInFlightBytes -= bytesToAppend.length;
-                            actor.run(this::peekBlock);
-                          });
+                      // happy path
+                      appendEntryLimiter.onCommit(lastEventPosition);
+                      currentInFlightBytes -= bytesToAppend.length;
+                      actor.run(this::peekBlock);
                     }
                   });
-        });
+            });
   }
 
   /* Iterate over the events in buffer and find the position of the last event */
@@ -181,7 +198,35 @@ public class LogStorageAppender extends Actor {
 
   @Override
   protected void onActorStarting() {
+    final CompletableActorFuture<Long> appendIndexFuture = new CompletableActorFuture<>();
+    distributedLog
+        .getLastAppendIndex()
+        .thenAccept(appendIndexFuture::complete)
+        .exceptionally(
+            t -> {
+              appendIndexFuture.completeExceptionally(t);
+              return null;
+            });
 
+    actor.runOnCompletionBlockingCurrentPhase(
+        appendIndexFuture,
+        (v, t) -> {
+          if (t == null) {
+            appendIndex = v;
+          } else {
+            LOG.error(
+                "Failed to retrieve last append index for partition {}. Close log appender.",
+                partitionId,
+                t);
+            // rethrow to fail actor starting
+            throw new UncheckedExecutionException(
+                "Failed to retrieve last append index for partition " + partitionId, t);
+          }
+        });
+  }
+
+  @Override
+  protected void onActorStarted() {
     actor.consume(writeBufferSubscription, this::peekBlock);
   }
 

--- a/logstreams/src/test/java/io/zeebe/distributedlog/DistributedLogPartitionRule.java
+++ b/logstreams/src/test/java/io/zeebe/distributedlog/DistributedLogPartitionRule.java
@@ -12,6 +12,7 @@ import static io.zeebe.logstreams.impl.service.LogStreamServiceNames.logStreamSe
 import static io.zeebe.util.buffer.BufferUtil.bufferAsString;
 import static io.zeebe.util.buffer.BufferUtil.wrapString;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.spy;
 
 import io.zeebe.distributedlog.impl.DistributedLogstreamPartition;
 import io.zeebe.logstreams.log.BufferedLogStreamReader;
@@ -44,6 +45,7 @@ public class DistributedLogPartitionRule {
   private final String logName;
   private LogStream logStream;
   private BufferedLogStreamReader reader;
+  private DistributedLogstreamPartition currentLogSpy;
 
   public DistributedLogPartitionRule(
       ServiceContainer serviceContainer, int nodeId, int partition, Path rootDirectory)
@@ -80,12 +82,11 @@ public class DistributedLogPartitionRule {
   }
 
   private void createDistributedLog() {
-    final DistributedLogstreamPartition log =
-        new DistributedLogstreamPartition(partition, leaderTerm++);
+    currentLogSpy = spy(new DistributedLogstreamPartition(partition, leaderTerm++));
     final ActorFuture<DistributedLogstreamPartition> installFuture =
         serviceContainer
-            .createService(distributedLogPartitionServiceName(logName), log)
-            .dependency(DistributedLogRule.ATOMIX_SERVICE_NAME, log.getAtomixInjector())
+            .createService(distributedLogPartitionServiceName(logName), currentLogSpy)
+            .dependency(DistributedLogRule.ATOMIX_SERVICE_NAME, currentLogSpy.getAtomixInjector())
             .install();
     installFuture.join();
   }
@@ -96,6 +97,10 @@ public class DistributedLogPartitionRule {
     }
 
     logStream.openAppender().join();
+  }
+
+  public DistributedLogstreamPartition getCurrentLogSpy() {
+    return currentLogSpy;
   }
 
   public void becomeFollower() {
@@ -145,5 +150,30 @@ public class DistributedLogPartitionRule {
       numEvents++;
     }
     return numEvents;
+  }
+
+  public boolean eventsInOrder(Event[] events) {
+    if (events == null || events.length == 0) {
+      return true;
+    }
+
+    reader.seek(events[0].position);
+    for (Event event : events) {
+      if (reader.hasNext()) {
+        final LoggedEvent loggedEvent = reader.next();
+        final String messageRead =
+            bufferAsString(
+                loggedEvent.getValueBuffer(),
+                loggedEvent.getValueOffset(),
+                loggedEvent.getValueLength());
+        final long eventPosition = loggedEvent.getPosition();
+        if (!event.message.equals(messageRead) || eventPosition != event.position) {
+          return false;
+        }
+      } else {
+        return false;
+      }
+    }
+    return true;
   }
 }

--- a/logstreams/src/test/java/io/zeebe/distributedlog/DistributedLogRule.java
+++ b/logstreams/src/test/java/io/zeebe/distributedlog/DistributedLogRule.java
@@ -19,6 +19,7 @@ import io.atomix.protocols.raft.partition.RaftPartitionGroup;
 import io.atomix.utils.net.Address;
 import io.zeebe.distributedlog.impl.DistributedLogstreamConfig;
 import io.zeebe.distributedlog.impl.DistributedLogstreamName;
+import io.zeebe.distributedlog.impl.DistributedLogstreamPartition;
 import io.zeebe.distributedlog.impl.LogstreamConfig;
 import io.zeebe.servicecontainer.ServiceContainer;
 import io.zeebe.servicecontainer.ServiceName;
@@ -208,6 +209,10 @@ public class DistributedLogRule extends ExternalResource {
     partitions.get(partitionId).becomeLeader();
   }
 
+  public DistributedLogstreamPartition getSpy(int partition) {
+    return partitions.get(partition).getCurrentLogSpy();
+  }
+
   public void becomeFollower(final int partitionId) {
     partitions.get(partitionId).becomeFollower();
   }
@@ -230,5 +235,9 @@ public class DistributedLogRule extends ExternalResource {
 
   public int getCommittedEventsCount(final int partitionId) {
     return partitions.get(partitionId).getCommittedEventsCount();
+  }
+
+  public boolean eventsInOrderAppended(int partitionId, Event[] events) {
+    return partitions.get(partitionId).eventsInOrder(events);
   }
 }

--- a/logstreams/src/test/java/io/zeebe/distributedlog/DistributedLogTest.java
+++ b/logstreams/src/test/java/io/zeebe/distributedlog/DistributedLogTest.java
@@ -9,6 +9,11 @@ package io.zeebe.distributedlog;
 
 import static io.zeebe.distributedlog.DistributedLogRule.LOG;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.doReturn;
 
 import io.zeebe.protocol.Protocol;
 import io.zeebe.servicecontainer.testing.ServiceContainerRule;
@@ -18,6 +23,8 @@ import java.io.IOException;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeoutException;
 import org.junit.Rule;
@@ -103,6 +110,52 @@ public class DistributedLogTest {
     assertEventsCount(node1, 3);
     assertEventsCount(node2, 3);
     assertEventsCount(node3, 3);
+  }
+
+  @Test
+  public void shouldReplicateMultipleEventsWithFailure()
+      throws ExecutionException, InterruptedException, TimeoutException {
+    // given
+    final CountDownLatch latch = new CountDownLatch(1);
+    final var failingFuture = new CompletableFuture<Long>();
+    node1.waitUntilNodesJoined();
+    node2.waitUntilNodesJoined();
+    node3.waitUntilNodesJoined();
+
+    node1.becomeLeader(PARTITION_ID);
+
+    doReturn(failingFuture)
+        .doCallRealMethod()
+        .when(node1.getSpy(PARTITION_ID))
+        .asyncAppend(eq(2L), any(byte[].class), anyLong());
+
+    doAnswer(
+            i -> {
+              latch.countDown();
+              return i.callRealMethod();
+            })
+        .doCallRealMethod()
+        .when(node1.getSpy(PARTITION_ID))
+        .asyncAppend(eq(3L), any(byte[].class), anyLong());
+
+    // when
+    final Event event1 = writeEvent("record1");
+    Thread.sleep(100L); // to get events in different chunks
+    final Event event2 = writeEvent("record2");
+    Thread.sleep(100L); // to get events in different chunks
+    final Event event3 = writeEvent("record3");
+    final Event event4 = writeEvent("record4");
+    latch.countDown();
+    failingFuture.completeExceptionally(new RuntimeException("Second event failed!"));
+
+    // then
+    assertEventsInOrderOnNode(node1, event1, event2, event3, event4);
+    assertEventsInOrderOnNode(node2, event1, event2, event3, event4);
+    assertEventsInOrderOnNode(node3, event1, event2, event3, event4);
+
+    assertEventsCount(node1, 4);
+    assertEventsCount(node2, 4);
+    assertEventsCount(node3, 4);
   }
 
   @Test
@@ -225,8 +278,7 @@ public class DistributedLogTest {
     assertThat(node.getCommittedEventsCount(PARTITION_ID)).isEqualTo(expectedCount);
   }
 
-  private class Event {
-    String message;
-    long position;
+  private void assertEventsInOrderOnNode(DistributedLogRule node, Event... events) {
+    TestUtil.waitUntil(() -> node.eventsInOrderAppended(PARTITION_ID, events), 10);
   }
 }

--- a/logstreams/src/test/java/io/zeebe/distributedlog/Event.java
+++ b/logstreams/src/test/java/io/zeebe/distributedlog/Event.java
@@ -1,0 +1,13 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.distributedlog;
+
+class Event {
+  String message;
+  long position;
+}

--- a/logstreams/src/test/java/io/zeebe/logstreams/LogStorageAppenderTest.java
+++ b/logstreams/src/test/java/io/zeebe/logstreams/LogStorageAppenderTest.java
@@ -1,0 +1,196 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.logstreams;
+
+import static io.zeebe.util.buffer.BufferUtil.wrapString;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.timeout;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import io.zeebe.dispatcher.Dispatcher;
+import io.zeebe.dispatcher.Dispatchers;
+import io.zeebe.dispatcher.Subscription;
+import io.zeebe.distributedlog.impl.DistributedLogstreamPartition;
+import io.zeebe.logstreams.impl.LogStorageAppender;
+import io.zeebe.logstreams.log.LogStream;
+import io.zeebe.logstreams.log.LogStreamWriterImpl;
+import io.zeebe.protocol.impl.record.RecordMetadata;
+import io.zeebe.test.util.TestUtil;
+import io.zeebe.util.ByteValue;
+import io.zeebe.util.exception.UncheckedExecutionException;
+import io.zeebe.util.sched.testing.ActorSchedulerRule;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicLong;
+import org.agrona.DirectBuffer;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.InOrder;
+import org.mockito.Mockito;
+
+public class LogStorageAppenderTest {
+  private static ByteValue maxFragmentSize = ByteValue.ofMegabytes(4);
+
+  @Rule public ActorSchedulerRule schedulerRule = new ActorSchedulerRule();
+
+  private DistributedLogstreamPartition primitiveMock;
+  private LogStorageAppender appender;
+  private LogStreamWriterImpl writer;
+
+  @Before
+  public void setup() {
+    primitiveMock = mock(DistributedLogstreamPartition.class);
+
+    when(primitiveMock.getLastAppendIndex()).thenReturn(CompletableFuture.completedFuture(0L));
+    when(primitiveMock.asyncAppend(anyLong(), any(byte[].class), anyLong()))
+        .thenReturn(CompletableFuture.completedFuture(1L));
+
+    final Dispatcher dispatcher =
+        Dispatchers.create("dispatcher")
+            .maxFragmentLength(maxFragmentSize)
+            .initialPartitionId(1)
+            .actorScheduler(schedulerRule.get())
+            .build();
+    final Subscription subscription = dispatcher.openSubscription("subscription");
+    appender =
+        new LogStorageAppender(
+            "appender", primitiveMock, subscription, (int) maxFragmentSize.toBytes());
+
+    final LogStream logStream = mock(LogStream.class);
+    when(logStream.getPartitionId()).thenReturn(1);
+    when(logStream.getWriteBuffer()).thenReturn(dispatcher);
+    writer = new LogStreamWriterImpl(logStream);
+  }
+
+  public long writeEvent(final String message) {
+    final AtomicLong writePosition = new AtomicLong();
+    final RecordMetadata metadata = new RecordMetadata();
+    final DirectBuffer value = wrapString(message);
+
+    TestUtil.doRepeatedly(
+            () -> writer.key(-1).metadataWriter(metadata.reset()).value(value).tryWrite())
+        .until(
+            position -> {
+              if (position != null && position >= 0) {
+                writePosition.set(position);
+                return true;
+              } else {
+                return false;
+              }
+            },
+            "Failed to write event with message {}",
+            message);
+    return writePosition.get();
+  }
+
+  @Test
+  public void shouldAppend() {
+    // given
+    schedulerRule.submitActor(appender).join();
+
+    // when
+    writeEvent("msg");
+
+    // then
+    verify(primitiveMock, timeout(1000)).asyncAppend(eq(1L), any(byte[].class), anyLong());
+  }
+
+  @Test
+  public void shouldAppendMultipleIncrementIndex() {
+    // given
+    schedulerRule.submitActor(appender).join();
+    writeEvent("msg");
+
+    // when
+    writeEvent("msg");
+
+    // then
+    final InOrder inOrder = Mockito.inOrder(primitiveMock);
+    inOrder.verify(primitiveMock, timeout(1000)).asyncAppend(eq(1L), any(byte[].class), anyLong());
+    inOrder.verify(primitiveMock, timeout(1000)).asyncAppend(eq(2L), any(byte[].class), anyLong());
+    inOrder.verifyNoMoreInteractions();
+  }
+
+  @Test
+  public void shouldStartWithIndex() {
+    // given
+    when(primitiveMock.getLastAppendIndex()).thenReturn(CompletableFuture.completedFuture(12L));
+    schedulerRule.submitActor(appender).join();
+
+    // when
+    writeEvent("msg");
+
+    // then
+    verify(primitiveMock, timeout(1000)).asyncAppend(eq(13L), any(byte[].class), anyLong());
+  }
+
+  @Test
+  public void shouldFailToStartOnRequestLastIndexError() {
+    // given
+    when(primitiveMock.getLastAppendIndex())
+        .thenReturn(CompletableFuture.failedFuture(new RuntimeException("failed")));
+
+    // when - expect exception
+    assertThatThrownBy(() -> schedulerRule.submitActor(appender).join())
+        .isInstanceOf(ExecutionException.class)
+        .hasCauseInstanceOf(UncheckedExecutionException.class);
+  }
+
+  @Test
+  public void shouldRetryOnError() {
+    // given
+    when(primitiveMock.asyncAppend(anyLong(), any(byte[].class), anyLong()))
+        .thenReturn(CompletableFuture.failedFuture(new RuntimeException("Failed to append!")))
+        .thenReturn(CompletableFuture.completedFuture(1L));
+    schedulerRule.submitActor(appender).join();
+
+    // when
+    writeEvent("msg");
+
+    // then
+    verify(primitiveMock, timeout(1000).times(2)).asyncAppend(eq(1L), any(byte[].class), anyLong());
+  }
+
+  @Test
+  public void shouldRetryOnFailedAppend() {
+    // given
+    when(primitiveMock.asyncAppend(anyLong(), any(byte[].class), anyLong()))
+        .thenReturn(CompletableFuture.completedFuture(-1L))
+        .thenReturn(CompletableFuture.completedFuture(1L));
+    schedulerRule.submitActor(appender).join();
+
+    // when
+    writeEvent("msg");
+
+    // then
+    verify(primitiveMock, timeout(1000).times(2)).asyncAppend(eq(1L), any(byte[].class), anyLong());
+  }
+
+  @Test
+  public void shouldRetryOnBothErrorTypes() {
+    // given
+    when(primitiveMock.asyncAppend(anyLong(), any(byte[].class), anyLong()))
+        .thenReturn(CompletableFuture.failedFuture(new RuntimeException("Failed to append!")))
+        .thenReturn(CompletableFuture.completedFuture(-1L))
+        .thenReturn(CompletableFuture.completedFuture(1L));
+
+    schedulerRule.submitActor(appender).join();
+
+    // when
+    writeEvent("msg");
+
+    // then
+    verify(primitiveMock, timeout(1000).times(3)).asyncAppend(eq(1L), any(byte[].class), anyLong());
+  }
+}

--- a/logstreams/src/test/java/io/zeebe/logstreams/util/LogStreamRule.java
+++ b/logstreams/src/test/java/io/zeebe/logstreams/util/LogStreamRule.java
@@ -11,6 +11,7 @@ import static io.zeebe.logstreams.impl.service.LogStreamServiceNames.distributed
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyLong;
 import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.doReturn;
 import static org.mockito.Mockito.mock;
 
 import io.zeebe.distributedlog.DistributedLogstreamService;
@@ -157,15 +158,18 @@ public final class LogStreamRule extends ExternalResource {
                       && arguments.length > 1
                       && arguments[0] != null
                       && arguments[1] != null) {
-                    final byte[] bytes = (byte[]) arguments[0];
-                    final long pos = (long) arguments[1];
+                    final long appendIndex = (long) arguments[0];
+                    final byte[] bytes = (byte[]) arguments[1];
+                    final long pos = (long) arguments[2];
                     return CompletableFuture.completedFuture(
-                        distributedLogImpl.append(nodeId, pos, bytes));
+                        distributedLogImpl.append(nodeId, appendIndex, pos, bytes));
                   }
                   return null;
                 })
         .when(mockDistLog)
-        .asyncAppend(any(), anyLong());
+        .asyncAppend(anyLong(), any(), anyLong());
+
+    doReturn(CompletableFuture.completedFuture(0L)).when(mockDistLog).getLastAppendIndex();
 
     serviceContainer
         .createService(distributedLogPartitionServiceName(builder.getLogName()), () -> mockDistLog)

--- a/logstreams/src/test/resources/log4j2-test.xml
+++ b/logstreams/src/test/resources/log4j2-test.xml
@@ -9,6 +9,7 @@
 
   <Loggers>
     <Logger name="io.zeebe" level="debug"/>
+    <Logger name="io.zeebe.logstreams" level="trace"/>
     <Logger name="io.atomix" level="info"/>
 
     <Root level="info">

--- a/util/src/main/java/io/zeebe/util/exception/UncheckedExecutionException.java
+++ b/util/src/main/java/io/zeebe/util/exception/UncheckedExecutionException.java
@@ -1,0 +1,19 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.util.exception;
+
+/**
+ * Should be use to indicate an unexpected exception during execution. This exception extends {@link
+ * RuntimeException} to be unchecked.
+ */
+public class UncheckedExecutionException extends RuntimeException {
+
+  public UncheckedExecutionException(String message, Throwable cause) {
+    super(message, cause);
+  }
+}

